driver-sync/src/test/functional/com/mongodb/client/AbstractClientSideEncryptionKmsTlsTest.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.client;
+
+import com.mongodb.ClientEncryptionSettings;
+import com.mongodb.MongoClientException;
+import com.mongodb.client.model.vault.DataKeyOptions;
+import com.mongodb.client.vault.ClientEncryption;
+import org.bson.BsonDocument;
+import org.jetbrains.annotations.NotNull;
+import org.junit.jupiter.api.Test;
+
+import java.security.cert.CertificateException;
+import java.security.cert.CertificateExpiredException;
+import java.util.HashMap;
+import java.util.Map;
+
+import static com.mongodb.client.Fixture.getMongoClientSettings;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.fail;
+import static org.junit.jupiter.api.Assumptions.assumeTrue;
+
+public abstract class AbstractClientSideEncryptionKmsTlsTest {
+
+    enum TlsErrorType {
+        EXPIRED() {
+            @Override
+            Class<?> getExpectedExceptionClass() {
+                return CertificateExpiredException.class;
+            }
+
+            @Override
+            String getExpectedExcepionMessageSubstring() {
+                return "NotAfter";
+            }
+        },
+        INVALID_HOSTNAME() {
+            @Override
+            Class<?> getExpectedExceptionClass() {
+                return CertificateException.class;
+            }
+
+            @Override
+            String getExpectedExcepionMessageSubstring() {
+                return "No subject alternative names present";
+            }
+        };
+        abstract Class<?> getExpectedExceptionClass();
+
+        abstract String getExpectedExcepionMessageSubstring();
+    }
+
+    private static final TlsErrorType EXPECTED_KMS_TLS_ERROR;
+
+    static {
+        if (System.getProperties().containsKey("org.mongodb.test.kms.tls.error.type")) {
+            String value = System.getProperty("org.mongodb.test.kms.tls.error.type");
+            if (value.equals("expired")) {
+                EXPECTED_KMS_TLS_ERROR = TlsErrorType.EXPIRED;
+            } else if (value.equals("invalidHostname")) {
+                EXPECTED_KMS_TLS_ERROR = TlsErrorType.INVALID_HOSTNAME;
+            } else {
+                throw new UnsupportedOperationException("Unsupported value: " + value);
+            }
+        } else {
+            EXPECTED_KMS_TLS_ERROR = null;
+        }
+    }
+
+    @NotNull
```

Comment:
We have `com.mongodb.lang.NonNull`, yet we use `org.jetbrains.annotations.NotNull` here, and also on the overriding methods. This rises two questions:

1. Why do we use JetBrains's annotation instead of ours?
2. Why do we duplicate it on overriding methods? If a super method is not allowed to return `null`, then neither is an overriding method, even if it does not duplicate the annotation. Duplicating it implies that the overriding method could also have been alternatively annotated with `Nullable`, which is not the case.

Reply:
1. IntelliJ auto-completed for me and I missed that I pulled in the wrong annotation
2. I removed the duplicated annotations.  Agreed that they are not necessary.

Reply:
Done